optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/DefaultPartitionedSearchPhase.java

```diff
@@ -151,34 +144,61 @@ public void solve(DefaultSolverScope<Solution_> solverScope) {
                 phaseScope.setLastCompletedStepScope(stepScope);
             }
         } finally {
-            // If a partition thread throws an Exception, it is propagated here
-            // but the other partition threads won't finish any time soon, so we need to ask them to terminate
-            childThreadPlumbingTermination.terminateChildren();
+            if (Thread.interrupted()) {
+                // 1a. If current thread is interrupted, propagate interrupt signal to children by initiating
+                // abrupt shutdown.
+                executor.shutdownNow();
+            } else {
+                // 1b. Otherwise, initiate graceful shutdown of the executor. This allows partition solvers to finish
+                // solving upon detecting the termination issued in the next step (2). Shutting down the executor
+                // service is important because the JVM cannot exit until all nondaemon threads have terminated.
+                executor.shutdown();
+            }
+
+            // 2. In case on of the partition threads threw an Exception, it is propagated here
+            // but the other partition threads are not aware of the failure and may continue solving for a long time,
+            // so we need to ask them to terminate. In case no exception was thrown, this does nothing.
+            if (!childThreadPlumbingTermination.terminateChildren()) {
+                // This may happen as a result of the abrupt shutdown in step 1a.
+                logger.info("Children have already been interrupted.");
+            }
+
+            // 3. Finally, wait until the executor finishes shutting down
             try {
-                // First wait for solvers to terminate voluntarily (because we have just issued children termination)
-                if (!threadPoolExecutor.awaitTermination(1, TimeUnit.SECONDS)) {
+                final int awaitingSeconds = 1;
+                if (!executor.awaitTermination(awaitingSeconds, TimeUnit.SECONDS)) {
                     // Some solvers refused to complete. Busy threads will be interrupted in the finally block
-                    logger.warn("{}Partitioned Search threadPoolExecutor didn't terminate within timeout (1 second).",
-                            logIndentation);
+                    logger.warn("{}Partitioned Search threadPoolExecutor didn't terminate within timeout ({} second).",
+                                logIndentation,
+                                awaitingSeconds);
```

Comment:
Tests provide line coverage for this `if`'s block but they don't provide mutation coverage as it is impossible (hard) to detect logging statements. Should we throw an exception instead?